### PR TITLE
Issue 771 secondary broadcast

### LINKED ISSUE #771: Reintroduce secondary broadcast
This will fix a few hacks, especially with ProcessedVariable
Need this for #760

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Reformat broadcasting to make sure it's always done in the right order and get rid of hacks

Fixes #771 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/expression_tree/broadcasts.py [hunk 1]
```diff
@@ -33,72 +32,84 @@ def __init__(
...
-    def check_and_set_domain_and_broadcast_type(
-        self, child, broadcast_domain, broadcast_type
+    def _unary_simplify(self, simplified_child):
+        """ See :meth:`pybamm.UnaryOperator.simplify()`. """
+        return self._unary_new_copy(simplified_child)
+
+
+class PrimaryBroadcast(Broadcast):
+    """A node in the expression tree representing a primary broadcasting operator.
+    Broadcasts in a `primary` dimension only. That is, makes explicit copies
```
Comment (rtimms) [inline pybamm/expression_tree/broadcasts.py]: I'm not sure the "That is, makes explicit copies" is particularly clear?

File: pybamm/expression_tree/broadcasts.py [hunk 2]
```diff
@@ -33,72 +32,84 @@ def __init__(
...
-                """Broadcast type must be either: 'primary', 'secondary', or 'full' and
-                not {}""".format(
-                    broadcast_type
-                )
+        "See :meth:`Broadcast.check_and_set_domains`"
+        # Can only do primary broadcast from current collector to electrode or from
+        # electrode to particle
+        if child.domain == []:
+            pass
+        elif child.domain[0] in [
```
Comment (rtimms) [inline pybamm/expression_tree/broadcasts.py]: can we do a similar check that if the domain is current collector, the broadcast domain must be and electrode or separator?

File: pybamm/expression_tree/broadcasts.py [hunk 3]
```diff
@@ -107,14 +118,10 @@ def evaluate_for_shape(self):
...
-            return np.outer(child_eval, vec).reshape(-1, 1)
-        elif self.broadcast_type == "full":
-            return child_eval * vec
+        return np.outer(child_eval, vec).reshape(-1, 1)
 
 
-class PrimaryBroadcast(Broadcast):
+class SecondaryBroadcast(Broadcast):
     """A node in the expression tree representing a primary broadcasting operator.
     Broadcasts in a `primary` dimension only. That is, makes explicit copies
```
Comment (rtimms) [inline pybamm/expression_tree/broadcasts.py]: "`secondary` dimension only", and as above the last part of this isn't clear to me

File: pybamm/expression_tree/broadcasts.py [hunk 4]
```diff
@@ -131,15 +138,53 @@ class PrimaryBroadcast(Broadcast):
...
-    def _unary_simplify(self, child):
-        """ See :meth:`pybamm.UnaryOperator.simplify()`. """
-        return PrimaryBroadcast(child, self.broadcast_domain)
+    def check_and_set_domains(
+        self, child, broadcast_type, broadcast_domain, broadcast_auxiliary_domains
+    ):
+        "See :meth:`Broadcast.check_and_set_domains`"
+
+        # Can only do secondary broadcast from particle to electrode or from
+        # current collector to electrode
```
Comment (rtimms) [inline pybamm/expression_tree/broadcasts.py]: from electrode to current collector?

File: pybamm/expression_tree/symbol.py [hunk 5]
```diff
@@ -168,6 +168,43 @@ def domain(self, domain):
...
+    def copy_domains(self, symbol):
+        "Copy the domains from a given symbol, bypassing checks"
+        self._domain = symbol.domain
+        self._auxiliary_domains = symbol.auxiliary_domains
+
+    def clear_domains(self):
+        "Clear domains, bypassing checks"
+        self._domain = []
+        self._auxiliary_domains = {}
+
```
Comment (rtimms) [inline pybamm/expression_tree/symbol.py]: nice!

File: pybamm/processed_variable.py [hunk 6]
```diff
@@ -224,88 +224,48 @@ def initialise_2D(self):
...
-            "secondary"
-        ] == ["positive electrode"]:
-            x_sol = self.mesh["positive electrode"][0].nodes
-            r_nodes = self.mesh["positive particle"][0].nodes
-            r_edges = self.mesh["positive particle"][0].edges
-            set_up_r = True
+            self.first_dimension = "r"
+            self.second_dimension = "x"
+            self.r_sol = first_dim_pts
+            self.x_sol = second_dim_pts
```
Comment (rtimms) [inline pybamm/processed_variable.py]: nice job tidying this up!

File: pybamm/expression_tree/broadcasts.py (hunk 1 above)
Comment (tinosulzer) [reply pybamm/expression_tree/broadcasts.py]: oh yeah I got distracted while writing this 😅

**Comment by rtimms** [comment]
yep, this looks great thanks. the notebook is very clear!

**Pull request merged**